File: TEMP/AtmaFileSystem/AtmaFileSystemSpecification/FileNameSpecification.cs
using System;
using AtmaFileSystem;
using TddEbook.TddToolkit;
using Xunit;

namespace AtmaFileSystemSpecification
{
  public class FileNameSpecification
  {
    [Fact]
    public void ShouldNotAllowToBeCreatedWithNullValue()
    {
      Assert.Throws<ArgumentNullException>(() => FileName.Value(null));
    }

    [Fact]
    public void ShouldThrowExceptionWhenCreatedWithStringContainingMoreThanJustAFileName()
    {
      Assert.Throws<ArgumentException>(() => FileName.Value(@"c:\\lolek\\lolki2.txt"));
    }

    [Fact]
    public void ShouldReturnNonNullFileNameWhenCreatedWithWellFormedPathString()
    {
      Assert.NotNull(FileName.Value("lolki2.txt"));
    }

    [Fact]
    public void ShouldBehaveLikeValueObject()
    {
      XAssert.IsValue<FileName>();
    }

    [Fact]
    public void ShouldReturnTheStringItWasCreatedWithWhenConvertedToString()
    {
      //GIVEN
      var initialValue = Any.String();
      var path = new FileName(initialValue);

      //WHEN
      var convertedToString = path.ToString();

      //THEN
      Assert.Equal(initialValue, convertedToString);
    }

    [Fact]
    public void ShouldAllowGettingExtensionWhenItExists()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();
      var extensionString = "." + Any.String();
      var fileNameWithExtensionString = fileNameWithoutExtensionString + extensionString;

      var fileNameWithExtension = new FileName(fileNameWithExtensionString);

      //WHEN
      var maybeExtension = fileNameWithExtension.Extension();

      //THEN
      Assert.True(maybeExtension.Found);
      Assert.Equal(new FileExtension(extensionString), maybeExtension.Value());
    }

    [Fact]
    public void ShouldYieldNoExtensionWhenThePathHasNoExtension()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();

      var fileNameWithoutExtension = new FileName(fileNameWithoutExtensionString);

      //WHEN
      var maybeExtension = fileNameWithoutExtension.Extension();

      //THEN
      Assert.False(maybeExtension.Found);
      Assert.Throws<InvalidOperationException>(() => maybeExtension.Value());
    }

    [Fact]
    public void ShouldAllowAccessingFileNameWithoutExtension()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();
      var extensionString = "." + Any.String();
      var fileNameWithExtensionString = fileNameWithoutExtensionString + extensionString;

      var fileNameWithExtension = new FileName(fileNameWithExtensionString);

      //WHEN
      var fileNameWithoutExtension = fileNameWithExtension.WithoutExtension();

      //THEN
      Assert.Equal(new FileNameWithoutExtension(fileNameWithoutExtensionString), fileNameWithoutExtension);

    }



    /* 
    TODO GetFileNameWithoutExtension Returns the file name of the specified path string without the extension.
    TODO GetRandomFileName Returns a random folder name or file name.
    TODO GetTempFileName Creates a uniquely named, zero-byte temporary file on disk and returns the full path of that file.
    TODO GetTempPath Returns the path of the current user's temporary folder.
    TODO HasExtension  Determines whether a path includes a file name extension.
    TODO IsPathRooted
    */
  }
}

File: Java/path-values/src/test/java/logic/FileNameSpecification.java
package logic;

import autofixture.publicinterface.Any;
import lombok.val;
import org.testng.annotations.Test;

import static com.github.grzesiek_galezowski.test_environment.XAssertJConditions.corectlyImplementedEquality;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class FileNameSpecification {

    @Test
    public void shouldNotAllowToBeCreatedWithNullValue() {
        assertThatThrownBy(() -> FileName.from(null))
            .isInstanceOf(NullPointerException.class);
    }

    @Test
    public void shouldThrowExceptionWhenCreatedWithStringContainingMoreThanJustAFileName() {
        assertThatThrownBy(() ->
            FileName.from("c:\\lolek\\lolki2.txt"))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void shouldHaveProperEquality() {
        assertThat(FileName.class).has(corectlyImplementedEquality());
    }

    @Test
    public void shouldReturnTheStringItWasCreatedWithWhenConvertedToString() {
        //GIVEN
        val initialValue = Any.alphaString();
        val path = FileName.from(initialValue);

        //WHEN
        val convertedToString = path.toString();

        //THEN
        assertThat(convertedToString).isEqualTo(initialValue);
    }

    /*



    @Test
    public void ShouldAllowGettingExtensionWhenItExists()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();
      var extensionString = "." + Any.String();
      var fileNameWithExtensionString = fileNameWithoutExtensionString + extensionString;

      var fileNameWithExtension = FileName.Value(fileNameWithExtensionString);

      //WHEN
      var maybeExtension = fileNameWithExtension.Extension();

      //THEN
      Assert.True(maybeExtension.Found);
      Assert.Equal(FileExtension.Value(extensionString), maybeExtension.Value());
    }

    @Test
    public void ShouldYieldNoExtensionWhenThePathHasNoExtension()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();

      var fileNameWithoutExtension = FileName.Value(fileNameWithoutExtensionString);

      //WHEN
      var maybeExtension = fileNameWithoutExtension.Extension();

      //THEN
      Assert.False(maybeExtension.Found);
      Assert.Throws<InvalidOperationException>(() => maybeExtension.Value());
    }

    @Test
    public void ShouldAllowAccessingFileNameWithoutExtension()
    {
      //GIVEN
      var fileNameWithoutExtensionString = Any.String();
      var extensionString = "." + Any.String();
      var fileNameWithExtensionString = fileNameWithoutExtensionString + extensionString;

      var fileNameWithExtension = FileName.Value(fileNameWithExtensionString);

      //WHEN
      var fileNameWithoutExtension = fileNameWithExtension.WithoutExtension();

      //THEN
      Assert.Equal(FileNameWithoutExtension.Value(fileNameWithoutExtensionString), fileNameWithoutExtension);

    }

    @Test
    public void ShouldAllowChangingExtension()
    {
      //GIVEN
      var fileName = FileName.Value(@"file.txt");

      //WHEN
      FileName nameWithNewExtension = fileName.ChangeExtensionTo(FileExtension.Value(".doc"));

      //THEN
      Assert.Equal(@"file.doc", nameWithNewExtension.ToString());

    }

    @Test
    public void ShouldDetermineEqualityToAnotherInstanceUsingFileSystemComparisonRules()
    {
      //GIVEN
      var path1 = Any.Instance<FileName>();
      var path2 = Any.Instance<FileName>();
      var fileSystemComparisonRules = Substitute.For<FileSystemComparisonRules>();
      var comparisonResult = Any.Boolean();

      fileSystemComparisonRules
        .ArePathStringsEqual(path1.ToString(), path2.ToString())
        .Returns(comparisonResult);


      //WHEN
      var equality = path1.ShallowEquals(path2, fileSystemComparisonRules);

      //THEN
      XAssert.Equal(comparisonResult, equality);
    }
 */

}